File: test/cli-bridge-cards.test.ts
import { describe, expect, it } from 'vitest';
import {
  buildCliBridgeApprovalCard,
  buildCliBridgeAwayNoticeCard,
  buildCliBridgeQuestionCard,
  buildCliBridgeTaskCompletionCard,
  cliBridgeSettingsSection,
  CLI,
} from '../src/cli-bridge/cards';
import type { CliHookStatus } from '../src/cli-bridge/types';

function values(node: unknown, acc: string[] = []): string[] {
  if (Array.isArray(node)) node.forEach((n) => values(n, acc));
  else if (node && typeof node === 'object') {
    const o = node as Record<string, unknown>;
    const behaviors = o.behaviors;
    if (Array.isArray(behaviors)) {
      for (const b of behaviors) {
        const v = (b as { value?: { a?: string } }).value;
        if (v?.a) acc.push(v.a);
      }
    }
    for (const k of Object.keys(o)) values(o[k], acc);
  }
  return acc;
}

describe('cli bridge cards', () => {
  const statuses: Record<'claude' | 'codex', CliHookStatus> = {
    claude: { agent: 'claude', status: 'installed', details: [] },
    codex: { agent: 'codex', status: 'conflict_agent2lark', details: ['old hook'] },
  };

  it('renders compact settings controls', () => {
    const section = cliBridgeSettingsSection({
      enabled: true,
      statuses,
      canEnable: { ok: true },
      notifyScope: 'all',
      agents: { claude: true, codex: true },
      keepAwake: true,
    });
    const json = JSON.stringify(section);
    expect(json).toContain('☕ 咖啡一下'); // recognizable feature name
    expect(json).toContain('咖啡一下：开'); // master toggle label
    // away copy now names lock as an instant-away signal
    expect(json).toContain('锁屏');
    expect(json).not.toContain('默认：仅当 Mac 空闲超过一段时间时转发');
    expect(json).toContain('与 agent2lark 冲突');
    // conflict present → warn that 修复 will overwrite agent2lark's hooks
    expect(json).toContain('会用本 bridge 覆盖它');
    // hooks are machine-global; surfaced so multi-bot users know it won't double-install
    expect(json).toContain('修复不会重复安装');
    expect(json).not.toContain('转发模式');
    expect(json).not.toContain('包含群聊会话用于调试');
    expect(values(section)).not.toContain(CLI.setDelivery);
    expect(values(section)).not.toContain(CLI.toggleIncludeBridge);
    expect(values(section)).toContain(CLI.repairHooks);
    // the three new axes each render their controls
    expect(json).toContain('📣 通知范围');
    expect(json).toContain('🤖 转发哪些后端');
    expect(json).toContain('🔋 离开保活');
    expect(values(section)).toContain(CLI.setNotifyScope);
    expect(values(section)).toContain(CLI.toggleAgent);
    expect(values(section)).toContain(CLI.toggleKeepAwake);
  });

  it('highlights the active notify scope and reflects per-agent / keep-awake state', () => {
    const section = cliBridgeSettingsSection({
      enabled: true,
      statuses,
      canEnable: { ok: true },
      notifyScope: 'bound_projects',
      agents: { claude: true, codex: false },
      keepAwake: false,
    });
    const json = JSON.stringify(section);
    expect(json).toContain('仅绑定项目');
    expect(json).toContain('Claude Code：开');
    expect(json).toContain('Codex：关');
    expect(json).toContain('离开保活：关');
    expect(values(section)).toContain(CLI.toggleAgent);
  });

  it('omits the agent2lark overwrite warning when there is no conflict', () => {
    const section = cliBridgeSettingsSection({
      enabled: true,
      statuses: {
        claude: { agent: 'claude', status: 'installed', details: [] },
        codex: { agent: 'codex', status: 'not_installed', details: [] },
      },
      canEnable: { ok: true },
      notifyScope: 'all',
      agents: { claude: true, codex: true },
      keepAwake: true,
    });
    expect(JSON.stringify(section)).not.toContain('会用本 bridge 覆盖它');
  });

  it('renders the away heads-up card (brand, agent, full cwd — the only card with cwd)', () => {
    const json = JSON.stringify(buildCliBridgeAwayNoticeCard({ source: 'claude', cwd: '/repo/x' }));
    expect(json).toContain('Vonvon Bridge');
    expect(json).toContain('当前项目'); // cwd section, unique to the away card
    expect(json).toContain('Claude Code');
    expect(json).toContain('/repo/x');
    // purely informational — carries no action callbacks
    expect(values(buildCliBridgeAwayNoticeCard({ source: 'codex', cwd: '/repo' }))).toHaveLength(0);
  });

  it('renders approval, question, and completion actions', () => {
    const approval = buildCliBridgeApprovalCard({
      id: 'p1',
      source: 'codex',
      cwd: '/repo',
      toolName: 'Bash',
      command: 'git status',
      hookEventName: 'PermissionRequest',
      sessionId: 'session-1',
    });
    const approvalJson = JSON.stringify(approval);
    expect(approvalJson).toContain('Vonvon Bridge');
    expect(approvalJson).toContain('Bash'); // tool chip in the meta line
    expect(approvalJson).toContain('💻 **命令**');
    expect(approvalJson).toContain('git status');
    expect(approvalJson).not.toContain('工作目录'); // cwd dropped on non-away cards
    expect(values(approval)).toEqual([
      CLI.approveOnce,
      CLI.approveSession,
      CLI.deny,
    ]);

    const question = buildCliBridgeQuestionCard({
      id: 'q1',
      source: 'claude',
      cwd: '/repo',
      questions: [{ question: 'Pick?', header: 'Choice', multiSelect: false, options: [{ label: 'A', description: 'Alpha' }, { label: 'B' }] }],
      hookEventName: 'PermissionRequest',
    });
    const questionJson = JSON.stringify(question);
    expect(questionJson).toContain('Vonvon Bridge');
    expect(questionJson).toContain('Pick?');
    expect(questionJson).toContain('select_static'); // single-select → dropdown
    expect(questionJson).toContain('A — Alpha'); // option label carries a short description
    expect(questionJson).toContain('都不合适'); // always-visible custom free-text box
    // one atomic form submit — no per-option / per-custom buttons that lock the card
    expect(values(question)).toEqual([CLI.questionSubmit]);

    const completion = buildCliBridgeTaskCompletionCard({
      id: 't1',
      source: 'codex',
      cwd: '/repo',
      sessionId: 'session-stop',
      hookEventName: 'Stop',
      status: 'completed',
      summary: 'done',
      replyEnabled: true,
      replyExpiresAt: new Date('2026-06-12T11:02:00+08:00').getTime(),
    });
    const completionJson = JSON.stringify(completion);
    expect(completionJson).toContain('Vonvon Bridge');
    expect(completionJson).toContain('等待确认');
    expect(completionJson).toContain('📝 **Agent 输出**');
    expect(completionJson).toContain('有效期至');
    expect(values(completion)).toContain(CLI.taskCompletionDone);
  });

  it('renders resolved local-agent cards without live actions', () => {
    const approved = buildCliBridgeApprovalCard({
      id: 'p3',
      source: 'codex',
      cwd: '/repo',
      toolName: 'Bash',
      command: 'git status',
      status: 'approved',
      sessionId: 's',
    });
    expect(JSON.stringify(approved)).toContain('✅ 已允许');
    expect(values(approved)).toEqual([]);

    const selected = buildCliBridgeQuestionCard({
      id: 'q3',
      source: 'claude',
      cwd: '/repo',
      questions: [{ question: 'Pick?', header: 'Choice', multiSelect: false, options: [{ label: 'A' }, { label: 'B' }] }],
      status: 'approved',
      answers: { 'Pick?': 'A' },
    });
    expect(JSON.stringify(selected)).toContain('✅ 已回答');
    expect(JSON.stringify(selected)).toContain('A'); // the chosen answer is shown
    expect(values(selected)).toEqual([]); // resolved view carries no live actions

    const done = buildCliBridgeTaskCompletionCard({
      id: 't3',
      source: 'claude',
      cwd: '/repo',
      status: 'completed',
      summary: 'done',
      replyEnabled: false,
      replyDoneAt: new Date('2026-06-12T10:55:00+08:00').getTime(),
    });
    expect(JSON.stringify(done)).toContain('✅ 已确认完成');
    expect(JSON.stringify(done)).toContain('✅ 已完成');
    expect(values(done)).toEqual([]);
  });

  it('renders a multi-question form with single + multi-select dropdowns and per-question custom fields', () => {
    const card = buildCliBridgeQuestionCard({
      id: 'mq',
      source: 'claude',
      cwd: '/repo',
      questions: [
        { question: 'Deploy where?', header: 'Env', multiSelect: false, options: [{ label: 'staging' }, { label: 'prod' }] },
        { question: 'Which checks?', header: 'Checks', multiSelect: true, options: [{ label: 'lint' }, { label: 'test' }] },
      ],
    });
    const json = JSON.stringify(card);
    // questions are numbered only when there is more than one
    expect(json).toContain('1. Env');
    expect(json).toContain('2. Checks');
    expect(json).toContain('select_static'); // single-select → dropdown
    expect(json).toContain('multi_select_static'); // multi-select → multi dropdown
    expect(json).toContain('可多选');
    // every question carries its own always-visible custom free-text box
    expect(json).toContain('q0_custom');
    expect(json).toContain('q1_custom');
    // still exactly one atomic submit for the whole form
    expect(values(card)).toEqual([CLI.questionSubmit]);
  });

  it('hides session approval when allow-cache is disabled', () => {
    expect(values(buildCliBridgeApprovalCard({
      id: 'p2',
      source: 'codex',
      cwd: '/repo',
      toolName: 'Bash',
      allowSession: false,
    }))).toEqual([
      CLI.approveOnce,
      CLI.deny,
    ]);
  });

  it('caps an oversized final answer so the card stays within Feishu limits', () => {
    const huge = 'x'.repeat(50_000);
    const json = JSON.stringify(buildCliBridgeTaskCompletionCard({ id: 't2', source: 'claude', cwd: '/repo', status: 'completed', summary: huge, replyEnabled: true }));
    expect(json.length).toBeLessThan(10_000);
    expect(json).toContain('truncated');
  });

  it('inlines the local agents section into the global settings card (no sub-page)', async () => {
    const { buildSettingsCard } = await import('../src/card/dm-cards');
    const { cliBridgeSettingsSection } = await import('../src/cli-bridge/cards');
    const section = cliBridgeSettingsSection({
      enabled: true,
      statuses,
      canEnable: { ok: true },
      notifyScope: 'all',
      agents: { claude: true, codex: true },
      keepAwake: true,
    });
    const json = JSON.stringify(buildSettingsCard({
      accounts: { app: { id: 'app', secret: 'secret', tenant: 'feishu' } },
      preferences: { access: { ownerOpenId: 'ou_owner' } },
    }, section));
    // One card: the global settings header plus the inlined 咖啡一下 controls.
    expect(json).toContain('全局设置');
    expect(json).toContain('☕ 咖啡一下');
    expect(json).toContain('咖啡一下：开');
  });
});

File: internal/clibridge/clibridge_cards_test.go
package clibridge

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/modelzen/feishu-codex-bridge/internal/card"
)

func mustMarshal(t *testing.T, c card.CardObject) string {
	t.Helper()
	b, err := json.Marshal(c)
	if err != nil {
		t.Fatalf("marshal card: %v", err)
	}
	return string(b)
}

func TestBuildApprovalCard(t *testing.T) {
	c := BuildCliBridgeApprovalCard(struct {
		ID           string
		Source       CliBridgeAgent
		Cwd          string
		ToolName     string
		Command      string
		AllowSession bool
		Status       interactionStatus
		HookEventName string
		SessionID    string
		CreatedAt    int64
	}{
		ID: "p1", Source: AgentClaude, Cwd: "/proj", ToolName: "Bash",
		Command: "rm -rf /tmp/x", AllowSession: true,
	})
	s := mustMarshal(t, c)
	for _, want := range []string{CLI.ApproveOnce, CLI.Deny, "允许", "拒绝", "proj"} {
		if !strings.Contains(s, want) {
			t.Fatalf("approval card missing %q: %s", want, s)
		}
	}
	// AllowSession=true → 不应出现「始终允许」。
	if strings.Contains(s, CLI.ApproveSession) {
		t.Fatalf("allow-session should suppress 始终允许: %s", s)
	}
}

func TestBuildQuestionCard(t *testing.T) {
	c := BuildCliBridgeQuestionCard(struct {
		ID        string
		Source    CliBridgeAgent
		Cwd       string
		Questions []CliQuestionItem
		Status    interactionStatus
		Answers   map[string]string
		HookEventName string
		CreatedAt int64
	}{
		ID: "q1", Source: AgentClaude, Cwd: "/p",
		Questions: []CliQuestionItem{
			{Question: "Q1?", Options: []CliQuestionOption{{Label: "A"}, {Label: "B"}}},
		},
	})
	s := mustMarshal(t, c)
	if !strings.Contains(s, CLI.QuestionSubmit) {
		t.Fatalf("question card missing submit action: %s", s)
	}
	if !strings.Contains(s, QuestionChoiceField(0)) {
		t.Fatalf("question card missing choice field: %s", s)
	}
	if !strings.Contains(s, QuestionCustomField(0)) {
		t.Fatalf("question card missing custom field: %s", s)
	}
}

func TestBuildTaskCompletionCardReply(t *testing.T) {
	c := BuildCliBridgeTaskCompletionCard(struct {
		ID             string
		Source         CliBridgeAgent
		Cwd            string
		Status         string
		Summary        string
		ReplyEnabled   bool
		SessionID      string
		HookEventName  string
		CreatedAt      int64
		ReplyExpiresAt int64
		ReplyDoneAt    int64
	}{
		ID: "t1", Source: AgentClaude, Cwd: "/p", Status: "completed",
		Summary: "all done", ReplyEnabled: true,
	})
	s := mustMarshal(t, c)
	if !strings.Contains(s, CLI.TaskCompletionDone) {
		t.Fatalf("task card missing done action: %s", s)
	}
	if !strings.Contains(s, "all done") {
		t.Fatalf("task card missing summary: %s", s)
	}
}

func TestBuildSettingsSection(t *testing.T) {
	input := CliBridgeSettingsSectionInput{Enabled: true, NotifyScope: "all", CanEnable: true}
	input.Statuses = map[CliBridgeAgent]CliHookStatus{
		AgentClaude: {Status: HookInstalled},
		AgentCodex:  {Status: HookInstalled},
	}
	input.Agents.Claude = true
	input.Agents.Codex = true
	input.KeepAwake = true
	els := BuildCliBridgeSettingsSection(input)
	b, err := json.Marshal(els)
	if err != nil {
		t.Fatalf("marshal settings: %v", err)
	}
	s := string(b)
	for _, want := range []string{CLI.ToggleEnabled, CLI.SetNotifyScope, CLI.ToggleAgent, CLI.RepairHooks, "☕"} {
		if !strings.Contains(s, want) {
			t.Fatalf("settings missing %q: %s", want, s)
		}
	}
}

// TestShortElemFeishuConstraint 回归测试：飞书 CardKit 要求 element_id / form name
// ≤ 20 字符、仅 [a-zA-Z0-9_]、且以字母开头。shortElem 必须把长交互 ID 截断到合法范围。
func TestShortElemFeishuConstraint(t *testing.T) {
	longID := "d101771cde74f877b8da8bf2d712f5ae" // 真实场景里的 32 字符交互 ID
	cases := []struct {
		prefix string
		id     string
	}{
		{"approval_", longID},
		{"question_submit_", longID},
		{"cli_question_", longID},
		{"task_done_", longID},
		{"task_done_disabled_", longID},
		{"a", "weird id/with-dash.and.dots"},
	}
	for _, c := range cases {
		got := shortElem(c.prefix, c.id)
		if len(got) > 20 {
			t.Fatalf("shortElem(%q,%q)=%q 超过 20 字符", c.prefix, c.id, got)
		}
		if got == "" || (got[0] < 'a' || got[0] > 'z') && (got[0] < 'A' || got[0] > 'Z') {
			t.Fatalf("shortElem(%q,%q)=%q 必须以字母开头", c.prefix, c.id, got)
		}
		for _, r := range got {
			ok := (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') || r == '_'
			if !ok {
				t.Fatalf("shortElem(%q,%q)=%q 含非法字符 %q", c.prefix, c.id, got, string(r))
			}
		}
	}
}

// TestCardElementIDsWithinLimit 回归测试：用真实长度的交互 ID 渲染三类卡片，
// 断言 JSON 里每个 element_id / name 都 ≤ 20 字符（否则飞书会拒绝建卡）。
func TestCardElementIDsWithinLimit(t *testing.T) {
	longID := "d101771cde74f877b8da8bf2d712f5ae"
	cards := []card.CardObject{
		BuildCliBridgeApprovalCard(struct {
			ID           string
			Source       CliBridgeAgent
			Cwd          string
			ToolName     string
			Command      string
			AllowSession bool
			Status       interactionStatus
			HookEventName string
			SessionID    string
			CreatedAt    int64
		}{ID: longID, Source: AgentClaude, Cwd: "/p", ToolName: "Bash", Command: "rm -rf /x", AllowSession: false}),
		BuildCliBridgeQuestionCard(struct {
			ID        string
			Source    CliBridgeAgent
			Cwd       string
			Questions []CliQuestionItem
			Status    interactionStatus
			Answers   map[string]string
			HookEventName string
			CreatedAt int64
		}{ID: longID, Source: AgentClaude, Cwd: "/p", Questions: []CliQuestionItem{
			{Question: "Q?", Options: []CliQuestionOption{{Label: "A"}, {Label: "B"}}},
		}}),
		BuildCliBridgeTaskCompletionCard(struct {
			ID             string
			Source         CliBridgeAgent
			Cwd            string
			Status         string
			Summary        string
			ReplyEnabled   bool
			SessionID      string
			HookEventName  string
			CreatedAt      int64
			ReplyExpiresAt int64
			ReplyDoneAt    int64
		}{ID: longID, Source: AgentClaude, Cwd: "/p", Status: "completed", Summary: "ok", ReplyEnabled: true}),
	}
	for _, c := range cards {
		s := mustMarshal(t, c)
		var doc any
		if err := json.Unmarshal([]byte(s), &doc); err != nil {
			t.Fatalf("unmarshal card: %v", err)
		}
		walkAndCheckIDs(t, doc, "root")
	}
}

func walkAndCheckIDs(t *testing.T, v any, path string) {
	t.Helper()
	switch node := v.(type) {
	case map[string]any:
		for k, val := range node {
			if k == "element_id" || k == "name" {
				str, ok := val.(string)
				if ok && len(str) > 20 {
					t.Fatalf("element_id/name 超长 @ %s: %q (len=%d)", path, str, len(str))
				}
			}
			walkAndCheckIDs(t, val, path+"."+k)
		}
	case []any:
		for i, item := range node {
			walkAndCheckIDs(t, item, path+"["+itoa(i)+"]")
		}
	}
}
